services/data/src/react/hooks/stableValueHash.ts

```diff
@@ -0,0 +1,48 @@
+function hasObjectPrototype(o: any): boolean {
+    return Object.prototype.toString.call(o) === '[object Object]'
+}
+
+// eslint-disable-next-line @typescript-eslint/ban-types, @typescript-eslint/explicit-module-boundary-types
+export function isPlainObject(o: any): o is Object {
+    if (!hasObjectPrototype(o)) {
+        return false
+    }
+
+    // If has modified constructor
+    const ctor = o.constructor
+    if (typeof ctor === 'undefined') {
+        return true
+    }
+
+    // If has modified prototype
+    const prot = ctor.prototype
+    if (!hasObjectPrototype(prot)) {
+        return false
+    }
+
+    // If constructor does not have an Object-specific method
+    if (!Object.prototype.hasOwnProperty.call(prot, 'isPrototypeOf')) {
+        return false
+    }
+
+    // Most likely a plain Object
+    return true
+}
+
+/**
+ * Hashes the value into a stable hash.
+ */
+
+// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
+export function stableValueHash(value: any): string {
+    return JSON.stringify(value, (_, val) =>
+        isPlainObject(val)
+            ? Object.keys(val)
+                  .sort()
+                  .reduce((result, key) => {
+                      result[key] = val[key]
+                      return result
+                  }, {} as any)
+            : val
+    )
```

Comment:
I think that if this were to fail you'd be looking a very generic error message, i.e. `Uncaught TypeError: Converting circular structure to JSON`. As such I wonder if it would perhaps be better to wrap this in a try/catch block and throw a more descriptive error.

Could be there is another place in the code that already ensures this object is serialisable already, in which case you can forget about this remark.

Reply:
Yeah true. I don't think we check for circular references (or serializability) anywhere before this code. The third party lib I used in the first commit did handle (most) circular references by replacing them with a string.

We could deal with this more specifically. I do suspect that it's very rare for this to happen (and if you have this in your variables it's likely that ract-query's serialization, which this is based on, will fail before ours does).

Reply:
@ismay actually because the variables are just passed to the dynamic query arrow functions (i.e. `params: (variables) => ...something`) I don't think there's any strict requirement currently that they be serializable.  If switching to i.e. gql they would need to be

Reply:
The variables themselves never make it to react-query, right?  Just the "resolved" query object which calls all the dynamic query functions with the provided variables and expects a serializable output

Reply:
They do make it to react-query for serialization. React-query will serialize everything we pass to it as a 'dependency' for the current query (i.e. the query key), which includes the query and the variables. So:

1. If there's something in there that is not serializable, say a function, then that'll be ignored for the query key. Which should be fine I think.
2. If there's a circular reference then that'll crash react-query's serialization (I think, I haven't completely gone through their implementation).

So I'm not passing the resolved query, but instead the parts that contribute to the uniqueness of the query. Do you think it would be better to use the resolved query?

Reply:
I also wonder what usecase would use non-serializable variables btw, I can't really imagine it atm.

Reply:
> I also wonder what usecase would use non-serializable variables btw, I can't really imagine it atm.

I can't think of any vaild use-case either. The context I had in mind when making my suggestion was more in case a dev accidentally passed a non-serialisable object. Seeing a meaningful error message would be a good starting point for the bughunt.

Reply:
Yeah I'm not opposed to adding a more specific error. It won't necessarily fail when something can't be serialized though. But it'll definitely fail for an object with a circular reference. Maybe the error message should be, `Could not serialize variables. Make sure that the variables do not contain circular references and can be processed by JSON.stringify.`? Something like that?

Reply:
Proposed a fix here: https://github.com/dhis2/app-runtime/pull/949